### PR TITLE
Counting Jastrow Factors

### PR DESCRIPTION
**Pull request by bwvdg**
This is an implementation of Counting Jastrow factors, which are a type of four-body jastrow factor using a set of gaussian quotients as Jastrow basis functions. These basis functions depend on explicit electronic coordinates and are handled separately from the existing DistanceTable-based Jastrow framework. More information can be found at https://arxiv.org/pdf/1811.03191.pdf .

The Counting Jastrow class is separated into three layers: the outer layer (CountingJastrow.h) explicitly handles the log-linear Jastrow parameters, the middle layer is a wrapper for the gaussian quotient basis functions (CountingRegion.h), and the final layer contains the primitive basis function components (3d-gaussians) themselves (CountingFunctor.h). A builder class (CountingJastrowBuilder), a working test (short-H4-cjs-16-1), and a basic cjf generation utility (utils/ncjf_gen.py) are also included.

### CODE REVIEW

**Comment by prckent** [comment]
Many thanks for this!

You labeled the PR as a work in progress [WIP]. Do you plan any other updates?

**Comment by prckent** [comment]
One question:

In the test you have:
LIST(APPEND H4_CJS_SCALARS "totenergy" "-2.0772 0.0095") # total energy

Where did this reference energy and error bar come from? In our experience this needs to come from a run with (say) 100x more blocks in order to have reliable statistics. (We currently believe that most of the statistical failures on https://cdash.qmcpack.org/CDash/viewProjects.php are due to insufficiently accurate reference data due to overly short reference runs. They will be updated soon.)

**Comment by prckent** [comment]
Please could you remove the large sections of commented code? You can always add them back uncommented in a future update should they need to be activated.

**Comment by prckent** [comment]
We should aim to get this merged in before the new release, which we are working on now. The release will be followed by #794 which will require rebasing these changes.

**Comment by bwvdg** [comment]
I labelled this WIP because I was concerned about some code style (ex. using std::function as a getter/setter for data storage vectors). I don't plan on any substantial changes (such as those that might affect a test) any time soon.

I removed DiffCountingJastrowOrbital.h, which is unimplemented and should be the only source of the large code comments.

I took the energy and sigma directly from the test itself and should currently match the statistics calculated by check_scalar.py. I have numerical results for the same test under much better statistical conditions, but the test doesn't pass under the short conditions. The test currently consists of three linear method steps followed by a VMC run. I would like the test to include optimizer/evaluateDerivatives code, but I can't increase the sample size very much without making it take longer than about a minute. How should I modify the test?

**Comment by prckent** [comment]
A way to generate reliable reference data:
- Run a long vmc with x100 more blocks than the short test that can run in about a minute
- Compute average properties and error bars
- Put the computed average properties from above with appropriately rescaled error bars (x10 here) in the test definition

Averaged properties from shorter runs tend to be just bad enough that we see one or two sigma failure rates, and not the desired 3 sigma.

File: src/OhmmsData/libxmldefs.h [hunk 1]
```diff
@@ -70,7 +70,8 @@ bool putContent(T& a, xmlNodePtr cur)
   std::istringstream
   stream((const char*)(xmlNodeListGetString(cur->doc, cur->xmlChildrenNode, 1)));
   stream >> a;
-  return stream.good();
+  return !stream.fail();
```
Comment (ye-luo) [inline src/OhmmsData/libxmldefs.h]: Could you revert these changes or there is a reason? It seems irrelevant to this PR.

**Comment by prckent** [comment]
Revision of above statement: if you are testing optimization (vs a simple VMC run), we are open to ideas. We would like separate tests for optimization and for VMC. The latter we know how to make reliable enough.

File: src/QMCWaveFunctions/Jastrow/CountingFunctor.h [hunk 2]
```diff
@@ -0,0 +1,433 @@
+
+#ifndef QMCPLUSPLUS_COUNTING_FUNCTOR_H
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Jastrow/CountingFunctor.h]: Missing licence header. Please add and update the header at the beginning of every file you introduced.

File: src/QMCWaveFunctions/Jastrow/CountingFunctor.h [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+    {
+      os << std::endl;
+      for(int J = 0, IJ = I*3; J < 3; ++J, ++IJ)
+        os << boost::format("  %10.5f") % A[IJ];
+    }
+    os << std::endl << "  opt_A: ";
+    std::copy(opt_A.begin(), opt_A.end(), std::ostream_iterator<bool>(os,", "));
+    os << std::endl << "  B: "; 
+    for(auto it = B.begin(); it != B.end(); ++it)
+      os << boost::format("  %10.5f") % (*it);
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Jastrow/CountingFunctor.h]: Can we avoid boost here?

File: src/QMCWaveFunctions/Jastrow/CountingJastrowBuilder.cpp [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  {
+  ClassName="CountingJastrowBuilder";
+  NameOpt="0";
+  TypeOpt="unknown";
+  RegionOpt="normalized_gaussian";
+  SourceOpt=targetPtcl.getName();
+}
+
+template<class precision, template<class> class CountingRegionType>
+class CountingJastrowTypeHelper
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Jastrow/CountingJastrowBuilder.cpp]: Can we remove this class and directly put its content at its use? It seems it is only used once.

File: src/QMCWaveFunctions/Jastrow/CountingJastrowOrbital.h [hunk 5]
```diff
@@ -0,0 +1,741 @@
+#ifndef QMC_PLUS_PLUS_COUNTING_JASTROW_ORBITAL_H
+#define QMC_PLUS_PLUS_COUNTING_JASTROW_ORBITAL_H
+
+#include "Configuration.h"
+#include "Particle/ParticleSet.h"
+#include "QMCWaveFunctions/WaveFunctionComponent.h"
+
+#include "boost/format.hpp"
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Jastrow/CountingJastrowOrbital.h]: No boost please.

File: src/OhmmsData/libxmldefs.h (hunk 1 above)
Comment (bwvdg) [reply src/OhmmsData/libxmldefs.h]: The put function logic didn't consistently work without this, since stream.good() returns false if the stream reached eof.

Comment (ye-luo) [reply src/OhmmsData/libxmldefs.h]: Changing this may affect all other places using this function.
If xml input doesn't provide sufficient values to feed the request, it is a failure.
Please give your logic with an example.

**Comment by bwvdg** [comment]
I'll add the VMC test using the higher-resolution results.

I know in previous versions of QMCPACK, I've been able to get matching output down to machine precision even for multiple optimizer steps. For optimization, would it be unreasonable to test for consistency using a particular random seed (or even a small set of them) rather than for statistical accuracy? I do know this complicates future changes to walker loading/sampling or random number generation, but it seems more robust to optimizer divergences or small errors in a short statistical test, with the added benefit of not requiring a long sampling procedure.

**Comment by grahamlopez** [comment]
okay to test

File: src/OhmmsData/libxmldefs.h (hunk 1 above)
Comment (ye-luo) [reply src/OhmmsData/libxmldefs.h]: !stream.fail() is correct. Please remove all the commented stream.good().

File: src/QMCWaveFunctions/Jastrow/CountingJastrowOrbital.h [hunk 6]
```diff
@@ -0,0 +1,742 @@
...
+#include "Particle/ParticleSet.h"
+#include "QMCWaveFunctions/WaveFunctionComponent.h"
+#include "QMCWaveFunctions/Jastrow/CountingRegion.h"
+
+namespace qmcplusplus
+{
+
+template <class RegionType>
+class CountingJastrowOrbital: public WaveFunctionComponent 
+{
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Jastrow/CountingJastrowOrbital.h]: Could you change the class and file names to CountingJastrow. Calling it Orbital only adds confusion.

File: src/OhmmsData/libxmldefs.h (hunk 1 above)
Comment (ye-luo) [reply src/OhmmsData/libxmldefs.h]: I did it.

File: src/QMCWaveFunctions/Jastrow/CountingJastrowOrbital.h (hunk 6 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/Jastrow/CountingJastrowOrbital.h]: I changed it.

**Comment by ye-luo** [comment]
@bwvdg I fixed the complex compilation issue and CI builds all pass now. It is not intended to use this Jastrow with the complex code but to get the compilation through. If someone wants to run the complex code, careful check is still needed.

Please complete the following:
1) upload your run which generated the reference numbers. Add a folder qmc_cjs_ref under H4_ae and put **only** the xml input, standard output and *.scalar.dat 
2) about std::function. It seems you are building a matrix with vector storage. In QMCPACK, you can use the matrix class defined in OhmmsPETE/OhmmsMatrix.h. The matrix class accepts RealType/ValueType/PosType/GradType as its template argument.

**Comment by bwvdg** [comment]
Sorry for the delay. I moved the test to its own folder and replaced the indexing functions with Matrix classes.

**Comment by ye-luo** [comment]
@bwvdg I was not asking to move the tests but to add the input and output of the run you generates the reference data. The test and reference run should be different. See https://github.com/QMCPACK/qmcpack/wiki/How-to-add-tests.
You can take tests/molecules/LiH_dimer_pp as example.

**Comment by bwvdg** [comment]
Alright, I moved the test back. I have the input, scalarfile, and the output in the reference folder and the test in the original place.

**Comment by ye-luo** [comment]
@bwvdg The reference run should be much longer and wider than the test. You can't use the same one.

**Comment by bwvdg** [comment]
Wasn't thinking. I added the correct reference.

**Comment by ye-luo** [comment]
Please follow all the steps in
https://github.com/QMCPACK/qmcpack/wiki/How-to-add-tests
Write a README, update the test reference number based on the long reference run.
Once you think everything is ready to merge, remove the [WIP]

**Comment by bwvdg** [comment]
I added the readme and updated the reference number.

**Comment by ye-luo** [comment]
I revised the error bar based on the reference run uploaded by @bwvdg. The short test needs only 8 seconds on my workstation. There is no long test at the moment and we can add it after we observing the short one for a while.

**Comment by prckent** [comment]
Agree.

**Comment by prckent** [comment]
Many thanks for completing this @bwvdg
